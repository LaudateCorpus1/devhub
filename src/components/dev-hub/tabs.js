import React, { useState } from 'react';
import PropTypes from 'prop-types';
import ComponentFactory from '../ComponentFactory';
import { TabContext } from './tab-context';
import { getNestedValue } from '../../utils/get-nested-value';
import { Tabs as LeafyTabs, Tab } from '@leafygreen-ui/tabs';

export const SLUG_TO_STRING = {
    shell: 'Mongo Shell',
    compass: 'Compass',
    python: 'Python',
    javasync: 'Java (Sync)',
    'java-sync': 'Java (Sync)',
    nodejs: 'Node.js',
    php: 'PHP',
    motor: 'Motor',
    'java-async': 'Java (Async)',
    c: 'C',
    cpp: 'C++11',
    csharp: 'C#',
    perl: 'Perl',
    ruby: 'Ruby',
    scala: 'Scala',
    go: 'Go',
    cloud: 'Cloud',
    local: 'Local',
    macos: 'macOS',
    linux: 'Linux',
    windows: 'Windows',
    debian: 'Debian',
    rhel: 'RHEL',
};

const stringifyTab = tabName => {
    return SLUG_TO_STRING[tabName] || tabName;
};

const createFragment = (tab, index) => {
    return (
        <React.Fragment key={index}>
            {tab.children.map((tabChild, tabChildIndex) => (
                <ComponentFactory nodeData={tabChild} key={tabChildIndex} />
            ))}
        </React.Fragment>
    );
};

const Tabs = ({ nodeData }) => {
    const [activeTab, setActiveTab] = useState(0);

    const isHidden = nodeData.options && nodeData.options.hidden;
    const tabs = nodeData.children;

    return (
        <React.Fragment>
            {isHidden || (
                <LeafyTabs
                    selected={activeTab}
                    setSelected={setActiveTab}
                    darkMode
                >
                    {tabs.map((tab, index) => {
                        const tabId = getNestedValue(['options', 'tabid'], tab);
                        const tabTitle =
                            getNestedValue(['argument', 0, 'value'], tab) ||
                            stringifyTab(tabId);

                        // If there are no activeTabs, js would typically be disabled
                        const tabContent = createFragment(tab, index);

                        return (
                            <Tab
                                data-tabid={tabId}
                                role="tab"
                                key={index}
                                name={tabTitle}
                            >
                                {tabContent}
                            </Tab>
                        );
                    })}
                </LeafyTabs>
            )}
        </React.Fragment>
    );
};

export default Tabs;

Tabs.propTypes = {
    nodeData: PropTypes.shape({
        children: PropTypes.arrayOf(
            PropTypes.shape({
                argument: PropTypes.arrayOf(
                    PropTypes.shape({
                        value: PropTypes.string,
                    })
                ).isRequired,
                children: PropTypes.array,
                name: PropTypes.oneOf(['tab']),
                options: PropTypes.shape({
                    tabid: PropTypes.string.isRequired,
                }).isRequired,
            })
        ),
        options: PropTypes.shape({
            hidden: PropTypes.bool,
        }),
    }).isRequired,
};

Tabs.contextType = TabContext;
