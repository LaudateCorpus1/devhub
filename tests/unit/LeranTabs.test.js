import React from 'react';
import { shallow } from 'enzyme';
import Tabs from '../../src/components/dev-hub/learn-tabs';
import { Tab } from '@leafygreen-ui/tabs';

describe('Learn Tabs Component', () => {
    const mockTabsList = ['TabOne', 'TabTwo', 'TabThree'];
    const mockActiveItem = jest.fn();
    const mockHandleClick = jest.fn();
    const props = {
        tabsList: mockTabsList,
        handleClick: mockHandleClick,
        activeItem: mockActiveItem,
    };
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(<Tabs {...props} />);
    });

    test('renders component', () => {
        expect(wrapper).not.toBe(null);
    });

    test('has child component styled tabs', () => {
        expect(wrapper.find('StyledTabs')).not.toBe(null);
    });

    test('styled tabs has selected tab of 0 th element ', () => {
        expect(wrapper.find('StyledTabs')).not.toBe(null);
    });

    test('number of Tabs equal the number of elements passed from the tab list', () => {
        expect(wrapper.find(Tab)).toHaveLength(3);
    });

    test('order of tabs match the tab list', () => {
        expect(wrapper.find(Tab).at(0).props().name).toEqual('TabOne');
        expect(wrapper.find(Tab).at(1).props().name).toEqual('TabTwo');
        expect(wrapper.find(Tab).at(2).props().name).toEqual('TabThree');
    });
});
