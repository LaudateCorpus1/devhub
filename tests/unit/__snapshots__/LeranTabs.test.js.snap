// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Learn Tabs Component renders correctly 1`] = `
<Fragment>
  <StyledTabs
    as={[Function]}
    darkMode={true}
    data-test="tabs"
    selected={0}
    setSelected={[Function]}
  >
    <Tab
      data-tabid="TabOne"
      key="0"
      name="TabOne"
      role="tab"
    />
    <Tab
      data-tabid="TabTwo"
      key="1"
      name="TabTwo"
      role="tab"
    />
    <Tab
      data-tabid="TabThree"
      key="2"
      name="TabThree"
      role="tab"
    />
  </StyledTabs>
</Fragment>
`;
